File: library/src/main/java/cn/zfs/widget/BaseDialog.java
package cn.zfs.widget;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

/**
 * Created by zeng on 2016/12/4.
 */

public class BaseDialog implements DialogInterface {
    protected MyDialog dialog;
    protected Window window;
    protected View contentView;

    public BaseDialog(@NonNull Activity activity, @NonNull View view) {
        init(activity, view, 0);
    }

    public BaseDialog(@NonNull Activity activity, @LayoutRes int redId) {
        init(activity, View.inflate(activity, redId, null), 0);
    }

    public BaseDialog(@NonNull Activity context, @NonNull View view, @StyleRes int themeResId) {
        init(context, view, themeResId);
    }

    private void init(@NonNull Activity activity, @NonNull View view, @StyleRes int themeResId) {
        contentView = view;        
        dialog = new MyDialog(activity, themeResId);
        dialog.setOwnerActivity(activity);
        window = dialog.getWindow();
        window.requestFeature(Window.FEATURE_NO_TITLE);
        window.getDecorView().setPadding(0, 0, 0, 0);
        window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        dialog.setContentView(view);
    }

    public View getContentView() {
        return contentView;
    }
    
    public Context getContext() {
        return dialog.getContext();
    }

    public Activity getActivity() {
        return dialog.getOwnerActivity();
    }

    public <T extends View> T findViewById(@IdRes int id) {
        return contentView.findViewById(id);
    }

    public BaseDialog setCancelable(boolean cancelable) {
        dialog.setCancelable(cancelable);
        return this;
    }

    public BaseDialog setOnCancelListener(OnCancelListener listener) {
        dialog.setOnCancelListener(listener);
        return this;
    }

    public BaseDialog setOnDismissListener(OnDismissListener listener) {
        dialog.setOnDismissListener(listener);
        return this;
    }

    public BaseDialog setCanceledOnTouchOutside(boolean flag) {
        dialog.setCanceledOnTouchOutside(flag);
        return this;
    }

    public BaseDialog show() {
        if (!isDestroyed() && !dialog.isShowing()) {
            if (Looper.getMainLooper().equals(Looper.myLooper())) {
                dialog.show();
            } else {
                new Handler(Looper.getMainLooper()).post(new Runnable() {
                    @Override
                    public void run() {
                        dialog.show();
                    }
                });
            }
        }
        return this;
    }

    @Override
    public void cancel() {
        if (!isDestroyed()) {
            dialog.cancel();
        }
    }
    
    @Override
    public void dismiss() {
        if (!isDestroyed()) {
            dialog.dismiss();
        }        
    }

    /**
     * 设置返回键按下监听
     */
    public void setOnBackPressedListener(OnBackPressedListener listener) {
        dialog.setOnBackPressedListener(listener);
    }
    
    private boolean isDestroyed() {
        Activity activity = dialog.getOwnerActivity();
        return activity != null && (activity.isDestroyed() || activity.isFinishing());
    }
            
    public boolean isShowing() {
        return dialog.isShowing();
    }

    public BaseDialog setPadding(int left, int top, int right, int bottom) {
        window.getDecorView().setPadding(left, top, right, bottom);
        return this;
    }

    public BaseDialog setGravity(int g) {
        window.setGravity(g);
        return this;
    }

    public BaseDialog setAnimation(int rid) {
        window.setWindowAnimations(rid);
        return this;
    }

    /**
     * 设置偏移
     */
    public BaseDialog setOffset(int x, int y) {
        WindowManager.LayoutParams lp = window.getAttributes();
        lp.x = x;
        lp.y = y;
        window.setAttributes(lp);
        return this;
    }

    /**
     * 对话框外区域暗度
     */
    public BaseDialog setDimAmount(float amount) {
        window.setDimAmount(amount);
        return this;
    }
        
    public BaseDialog setSize(int width, int height) {
        WindowManager.LayoutParams lp = window.getAttributes();
        lp.width = width;
        lp.height = height;
        window.setAttributes(lp);
        return this;
    }

    /**
     * 旋转
     */
    public void setRotation(float rotation) {
        if (contentView != null) {
            contentView.setRotation(rotation);
        }
    }
    
    /**
     * 设置窗口类型，如{@link WindowManager.LayoutParams#TYPE_TOAST}设置一个window级别与toast相同的弹窗
     * @param type 窗口类型
     */
    public BaseDialog setWindowType(int type) {
        WindowManager.LayoutParams lp = window.getAttributes();
        lp.type = type;
        window.setAttributes(lp);
        return this;
    }
    
    public WindowManager.LayoutParams getAttributes() {
        return window.getAttributes();
    }
    
    public interface OnBackPressedListener {
        void onBackPressed();
    }
    
    public static class MyDialog extends Dialog {
        private OnBackPressedListener backPressedListener;
        
        MyDialog(@NonNull Context context) {
            super(context);
        }

        MyDialog(@NonNull Context context, int themeResId) {
            super(context, themeResId);
        }

        MyDialog(@NonNull Context context, boolean cancelable, @Nullable OnCancelListener cancelListener) {
            super(context, cancelable, cancelListener);
        }

        @Override
        public void onBackPressed() {
            super.onBackPressed();
            if (backPressedListener != null) {
                backPressedListener.onBackPressed();
            }
        }

        void setOnBackPressedListener(OnBackPressedListener backPressedListener) {
            this.backPressedListener = backPressedListener;
        }
    }
}

File: library/src/main/java/com/snail/widget/BaseDialog.kt
package com.snail.widget

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Looper
import android.support.annotation.IdRes
import android.support.annotation.LayoutRes
import android.support.annotation.StyleRes
import android.view.View
import android.view.Window
import android.view.WindowManager

/**
 * Created by zeng on 2016/12/4.
 */

open class BaseDialog : DialogInterface {
    protected lateinit var dialog: MyDialog
    protected var window: Window? = null
    lateinit var view: View
        protected set

    val context: Context
        get() = dialog.context

    val activity: Activity?
        get() = dialog.ownerActivity

    private val isDestroyed: Boolean
        get() {
            val activity = dialog.ownerActivity
            return activity != null && (activity.isDestroyed || activity.isFinishing)
        }

    val isShowing: Boolean
        get() = dialog.isShowing

    val attributes: WindowManager.LayoutParams
        get() = window!!.attributes

    constructor(activity: Activity, view: View) {
        init(activity, view, 0)
    }

    constructor(activity: Activity, @LayoutRes redId: Int) {
        init(activity, View.inflate(activity, redId, null), 0)
    }

    constructor(activity: Activity, view: View, @StyleRes themeResId: Int) {
        init(activity, view, themeResId)
    }

    private fun init(activity: Activity, view: View, @StyleRes themeResId: Int) {
        this.view = view
        dialog = MyDialog(activity, themeResId)
        dialog.ownerActivity = activity
        window = dialog.window
        window!!.requestFeature(Window.FEATURE_NO_TITLE)
        window!!.decorView.setPadding(0, 0, 0, 0)
        window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        dialog.setContentView(view)
    }

    fun <T : View> findViewById(@IdRes id: Int): T {
        return view.findViewById(id)
    }

    open fun setCancelable(cancelable: Boolean): BaseDialog {
        dialog.setCancelable(cancelable)
        return this
    }

    open fun setOnCancelListener(listener: DialogInterface.OnCancelListener): BaseDialog {
        dialog.setOnCancelListener(listener)
        return this
    }

    open fun setOnDismissListener(listener: DialogInterface.OnDismissListener): BaseDialog {
        dialog.setOnDismissListener(listener)
        return this
    }

    open fun setCanceledOnTouchOutside(flag: Boolean): BaseDialog {
        dialog.setCanceledOnTouchOutside(flag)
        return this
    }

    open fun show(): BaseDialog {
        if (!isDestroyed && !dialog.isShowing) {
            if (Looper.getMainLooper() == Looper.myLooper()) {
                dialog.show()
            } else {
                view.post { dialog.show() }
            }
        }
        return this
    }

    override fun cancel() {
        if (!isDestroyed) {
            dialog.cancel()
        }
    }

    override fun dismiss() {
        if (!isDestroyed) {
            dialog.dismiss()
        }
    }

    /**
     * 设置返回键按下监听
     */
    open fun setOnBackPressedListener(listener: OnBackPressedListener): BaseDialog {
        dialog.setOnBackPressedListener(listener)
        return this
    }

    open fun setPadding(left: Int, top: Int, right: Int, bottom: Int): BaseDialog {
        window!!.decorView.setPadding(left, top, right, bottom)
        return this
    }

    open fun setGravity(g: Int): BaseDialog {
        window!!.setGravity(g)
        return this
    }

    open fun setAnimation(rid: Int): BaseDialog {
        window!!.setWindowAnimations(rid)
        return this
    }

    /**
     * 设置偏移
     */
    open fun setOffset(x: Int, y: Int): BaseDialog {
        val lp = window!!.attributes
        lp.x = x
        lp.y = y
        window!!.attributes = lp
        return this
    }

    /**
     * 对话框外区域暗度
     */
    open fun setDimAmount(amount: Float): BaseDialog {
        window!!.setDimAmount(amount)
        return this
    }

    open fun setSize(width: Int, height: Int): BaseDialog {
        val lp = window!!.attributes
        lp.width = width
        lp.height = height
        window!!.attributes = lp
        return this
    }

    /**
     * 旋转
     */
    open fun setRotation(rotation: Float): BaseDialog {
        view.rotation = rotation
        return this
    }

    /**
     * 设置窗口类型，如[WindowManager.LayoutParams.TYPE_TOAST]设置一个window级别与toast相同的弹窗
     * @param type 窗口类型
     */
    open fun setWindowType(type: Int): BaseDialog {
        val lp = window!!.attributes
        lp.type = type
        window!!.attributes = lp
        return this
    }

    interface OnBackPressedListener {
        fun onBackPressed()
    }

    class MyDialog : Dialog {
        private var backPressedListener: OnBackPressedListener? = null

        internal constructor(context: Context) : super(context)

        internal constructor(context: Context, themeResId: Int) : super(context, themeResId)

        internal constructor(context: Context, cancelable: Boolean, cancelListener: DialogInterface.OnCancelListener?) : super(context, cancelable, cancelListener)

        override fun onBackPressed() {
            super.onBackPressed()
            if (backPressedListener != null) {
                backPressedListener!!.onBackPressed()
            }
        }

        internal fun setOnBackPressedListener(backPressedListener: OnBackPressedListener) {
            this.backPressedListener = backPressedListener
        }
    }
}